### ISSUE TITLE
The road to 1.0

### ISSUE THREAD

**Issue by emoon** [emoon/rust_minifb#101]
This crate has now been around for almost 4 years (Nov 22, 2015) so I guess it's time to start moving it towards 1.0 but before that there is some issues I want to be done before.

- [ ] #50 
- [ ] #47 
- [ ] #82 
- [ ] #95 
- [ ] #83

**Comment by tversteeg** [comment]
Maybe software support could be a feature?

**Comment by sinclairzx81** [comment]
Thanks for all your hard work on mini_fb, love this crate! Looking forward to the 1.0 release! :D

> Should it be possible to specific RGB(0) ordering of the buffers?

This would be a cool feature. I think currently, the buffers in mini_fb may linearly line up as BGRA (as little endian u32 arrays). Seeing as WASM is listed as an upcoming, it could be worth making note of the browsers `ImageData` using RGBA ordering as a pretty good rationale for including ordering as a feature.

>[ImageData](https://developer.mozilla.org/en-US/docs/Web/API/ImageData).data Read only
>> Is a Uint8ClampedArray representing a one-dimensional array containing the data in the RGBA order, with integer values between 0 and 255 (inclusive).

I wonder if having the option to define the order as well as being able to submit buffers as `Vec<u8>` might be possible. This may make it easier to transfer `ImageData` buffers between the Page and WASM context's without the need to remap the buffer prior to `update_with_buffer(...)`. With the `Vec<u8>` option added to help avoid the need to [transmute](https://github.com/sinclairzx81/black/blob/master/example/src/helpers/window.rs#L167-L178) between u8 and u32 buffers when setting RGBA values in the buffer. Also, `Vec<u8>` buffers makes the order super explicit when setting RGB values.

One common use case I expect would be to have the page load images (as RGBA) which are then passed to `mini_fb` to present (currently BGRA).

@tversteeg @FloVanGH Thoughts?

```rust
// possible Uint8ClampedArray passed from the page.
let mut buffer: Vec<u8> = vec![0; WIDTH * HEIGHT * 4];

let mut window = Window::new(
      "rgba ordering",
      WIDTH,
      HEIGHT,
      WindowOptions {
       order: Order::RGBA,
      ..WindowOptions::default()
    },
);
...
window.update_with_u8_buffer(&buffer).unwrap();
```
Again, thanks for all the work put into mini_fb!

**Comment by nicolasbauw** [comment]
I'm thinking about:
- 32-bit buffer as RGBA instead of 0RGB
- so, the buffer should adequately handle transparency to allow alpha compositing
- Option for the buffer swap to wait VBL for smoother animations (and avoid if we want to wait by counting time)
- accelerated blitting / clear area / clear whole buffer / texture generation / textured surfaces / scaling / rotation and so on functions

**Comment by emoon** [comment]
I think first 3 points are doable but I think that point 4 is out of scope for minifb.

**Comment by nifker** [comment]
@emoon What do you think about an event loop for 1.0?

**Comment by emoon** [comment]
Exactly how would that look like? I feel that the current design is simple and nice to use and I wouldn't like to break it if possible.

**Comment by nifker** [comment]
Similar to how SDL and GLFW do this.

**Comment by emoon** [comment]
GLFW uses callbacks and such for key presses and such. In general I want to avoid callbacks because it makes things more complicated. What are the biggest motivations for this change?

**Comment by nifker** [comment]
Im not talking about callbacks. GLFW uses an event loop: https://docs.rs/glfw/0.38.0/glfw/fn.flush_messages.html
We dont need to save all changes in specific structs so we basically dont even process them  but rather save the events the server sent and allow the developer to process them like they please to.

**Comment by emoon** [comment]
Alright. Yeah, that might be a good idea
